packages/cmr-client/src/CMR.ts

```diff
@@ -37,45 +29,91 @@ const userIpAddress = (): Promise<string> =>
-      token?: {
-        id: string
+    body : {
+      access_token: {
+        id: string,
+      },
+      expiration_date: {
+        id: BigInteger,
       }
     }
   };
   try {
-    response = await got.post(url, {
-      json: {
-        token: {
-          username: username,
-          password: password,
-          client_id: clientId,
-          user_ip_address: await userIpAddress(),
-          provider: cmrProvider,
+    response = await got.get(`https://${cmrenv}urs.earthdata.nasa.gov/api/users/tokens`,
+      {
+        headers: {
+          Authorization: `Basic ${buff}`,
         },
-      },
-      responseType: 'json',
-    });
+      }).json();
   } catch (error) {
-    if (get(error, 'response.body.errors')) {
-      throw new Error(`CMR Error: ${error.response.body.errors[0]}`);
+    logDetails.credentials = credentials;
+    log.error(error, logDetails);
+    const statusCode = get(error, 'response.statusCode', error.code);
+    const statusMessage = get(error, 'response.statusMessage', error.message);
+    let errorMessage = `Authentication error: Invalid Credentials, Authentication with Earthdata Login failed, statusCode: ${statusCode}, statusMessage: ${statusMessage}`;
+
+    const responseError = get(error, 'response.body.errors');
+    if (responseError) {
+      errorMessage = `${errorMessage}, CMR error message: ${JSON.stringify(responseError)}`;
     }
-    throw error;
-  }
 
-  if (!response.body.token) throw new Error('Authentication with CMR failed');
-
-  return response.body.token.id;
+    log.error(errorMessage);
+    throw new Error(errorMessage);
+  }
+  if (Object.keys(response).length === 0) {
+    let response2:
+    {
+      body: {
+        access_token?: {
+          id: string,
+        },
+        token_type?: {
+          id: string,
+        },
+        expiration_date?: {
+          id: BigInteger,
+        }
+      }
+    };
+    //response2: add a token to Earthdata if user credentials are valid
+    try {
+      response2 = await got.post(`https://${cmrenv}urs.earthdata.nasa.gov/api/users/token`,
+        {
+          headers: {
+            Authorization: `Basic ${buff}`,
+          },
+        }).json();
+    } catch (error) {
+      logDetails.credentials = credentials;
+      log.error(error, logDetails);
+      const statusCode = get(error, 'response.statusCode', error.code);
+      const statusMessage = get(error, 'response.statusMessage', error.message);
+      let errorMessage = `Authentication error: Invalid Credentials, Authentication with Earthdata Login failed, statusCode: ${statusCode}, statusMessage: ${statusMessage}`;
+      const responseError = get(error, 'response.body.errors');
+      if (responseError) {
+        errorMessage = `${errorMessage}, CMR error message: ${JSON.stringify(responseError)}`;
+      }
+      log.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+    return response2.access_token;
+  }
+  return response[0].access_token;
```

Comment:
only unexpired token should be retrieved

Reply:
It's easy to check if the token is still valid to avoid edge case, since it's in the return response.